### ISSUE TITLE
In-place modification detection does not work if a Tensor does not own its storage

### ISSUE THREAD

**Issue by patrick-kidger** [pytorch/pytorch#24413]
PyTorch checks if tensors have been modified in-place to ensure correctness of the backward operation, with `ctx.save_for_backward` and `ctx.saved_tensors` in the `forward` and `backward` methods of subclasses of `torch.autograd.Function`. This check does not appear to function when the saved tensor is not the 'owner' of its storage, e.g. when it is the result of a `view` or `transpose` operation.

To reproduce:
```python
import torch

class f(torch.autograd.Function):
    @staticmethod
    def forward(ctx, tensor1):
        # make sure we don't share storage with our input
        # (just to be clear that that's not what's going wrong)
        tensor2 = tensor1.clone()
        # tensor3 shares storage with tensor2
        # tensor3 = tensor2.view(-1) etc. would also demonstrate the problem
        tensor3 = tensor2.t()
        ctx.save_for_backward(tensor3)
        return tensor3
    @staticmethod
    def backward(ctx, grad):
        # trigger correctness check
        _ = ctx.saved_tensors
        return grad.t().clone()


x = torch.rand(2, 3, requires_grad=True)
y = f.apply(x)
y += 1
# Does not raise an error; this is the bug.
y.backward(torch.rand_like(y))


# In constrast, the following code correctly raises an error.
class ff(torch.autograd.Function):
    @staticmethod
    def forward(ctx, tensor1):
        tensor2 = tensor1.clone()
        ctx.save_for_backward(tensor2)
        return tensor2
    @staticmethod
    def backward(ctx, grad):
        _ = ctx.saved_tensors
        return grad.clone()

xx = torch.rand(2, 3, requires_grad=True)
yy = ff.apply(xx)
yy += 1
yy.backward(torch.rand_like(yy))
```

Expected behaviour is for an exception to be raised on the line `y.backward(torch.rand_like(y))`, in the same manner as an exception is raised on the line `yy.backward(torch.rand_like(yy))`.

 - PyTorch Version (e.g., 1.0): 1.2
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.7.4

cc @ezyang @SsnL @albanD

**Comment by soumith** [comment]
cc: @albanD @apaszke

**Comment by albanD** [comment]
Hi,

The problem is not with the inplace detection.
The problem here is that since you return a view from your Function, if you do an inplace operation on it, this inplace operation is attached to the origin of the view (and not your custom backward) and thus the `backward` that you define is never called.
Since the backward is not called, the saved variables are not unpacked and so the error you expect is not raised.

By looking at this quickly, maybe we don't want to detect the views when a variable does not require gradients? @apaszke

**Comment by ezyang** [comment]
I think this might be the same thing as #14707

**Comment by soumith** [comment]
this and other autograd tasks (especially high pri ones) would be great for Alban to tackle once he's back into the world next week :)

**Comment by albanD** [comment]
Closing this as a suplicate of https://github.com/pytorch/pytorch/issues/26546 
Currently working on this. This will soon raise an error as expected.
